teuthology/dispatcher/supervisor.py

```diff
@@ -0,0 +1,324 @@
+    job_info = serializer.job_info(job_config['name'], job_config['job_id'])
+    machine_status = query.get_statuses(job_info['targets'].keys())
+    # only unlock/nuke targets if locked in the first place
+    locked = [shortname(_['name'])
+              for _ in machine_status if _['locked']]
+    if not locked:
+        return
+    job_status = get_status(job_info)
+    if job_status == 'pass' or \
+            (job_config.get('unlock_on_failure', False) and not job_config.get('nuke-on-error', False)):
+        log.info('Unlocking machines...')
+        fake_ctx = create_fake_context(job_config)
+        for machine in locked:
+            teuthology.lock.ops.unlock_one(fake_ctx, machine, job_info['owner'],
+                                           job_info['archive_path'])
+    if job_status != 'pass' and job_config.get('nuke-on-error', False):
+        log.info('Nuking machines...')
+        fake_ctx = create_fake_context(job_config)
+        nuke(fake_ctx, True)
+
+
+def run_with_watchdog(process, job_config):
+    job_start_time = datetime.utcnow()
+
+    # Only push the information that's relevant to the watchdog, to save db
+    # load
+    job_info = dict(
+        name=job_config['name'],
+        job_id=job_config['job_id'],
+    )
+
+    # Sleep once outside of the loop to avoid double-posting jobs
+    time.sleep(teuth_config.watchdog_interval)
+    while process.poll() is None:
+        # Kill jobs that have been running longer than the global max
+        run_time = datetime.utcnow() - job_start_time
+        total_seconds = run_time.days * 60 * 60 * 24 + run_time.seconds
+        if total_seconds > teuth_config.max_job_time:
+            log.warning("Job ran longer than {max}s. Killing...".format(
+                max=teuth_config.max_job_time))
+            transfer_archives(job_info['name'], job_info['job_id'],
+                              teuth_config.archive_base, job_config)
+            kill_job(job_info['name'], job_info['job_id'],
+                     teuth_config.archive_base, job_config['owner'])
+
+        # calling this without a status just updates the jobs updated time
+        report.try_push_job_info(job_info)
+        time.sleep(teuth_config.watchdog_interval)
+
+    # we no longer support testing theses old branches
+    assert(job_config.get('teuthology_branch') not in ('argonaut', 'bobtail',
+                                                       'cuttlefish', 'dumpling'))
+
+    # Let's make sure that paddles knows the job is finished. We don't know
+    # the status, but if it was a pass or fail it will have already been
+    # reported to paddles. In that case paddles ignores the 'dead' status.
+    # If the job was killed, paddles will use the 'dead' status.
+    report.try_push_job_info(job_info, dict(status='dead'))
+
+
+def create_fake_context(job_config, block=False):
+    if job_config['owner'] is None:
+        job_config['owner'] = get_user()
```

Comment:
why are you mutating the `job_config`? do we expect a changed `job_config` after calling `create_fake_context()`?

Reply:
ah, my mistake. Reverted.